x-pack/plugins/cases/server/client/user_actions/connectors.ts

```diff
@@ -0,0 +1,244 @@
+  latestUserAction?: SavedObject<CaseUserActionResponse>;
+  actionsClient: PublicMethodsOf<ActionsClient>;
+  userActionService: CaseUserActionService;
+}): Promise<EnrichedConnector[]> => {
+  const pushInfo = await getPushInfo({ caseId, activity: connectors, userActionService });
+
+  const enrichedConnectors: EnrichedConnector[] = [];
+
+  for (const aggregationConnector of connectors) {
+    const connectorDetails = await actionsClient.get({ id: aggregationConnector.connectorId });
+    const connector = getConnectorInfoFromSavedObject(aggregationConnector.fields);
+
+    const latestUserActionCreatedAt = getDate(latestUserAction?.attributes.created_at);
+
+    if (connector != null) {
+      enrichedConnectors.push({
+        connector,
+        name: connectorDetails.name,
+        pushInfo: pushInfo.get(aggregationConnector.connectorId),
+        latestUserActionDate: latestUserActionCreatedAt,
+      });
+    }
+  }
+
+  return enrichedConnectors;
+};
+
+const getPushInfo = async ({
+  caseId,
+  activity,
+  userActionService,
+}: {
+  caseId: string;
+  activity: CaseConnectorActivity[];
+  userActionService: CaseUserActionService;
+}): Promise<Map<string, EnrichedPushInfo>> => {
+  const pushRequest: PushInfo[] = [];
+
+  for (const connectorInfo of activity) {
+    const pushCreatedAt = getDate(connectorInfo.push?.attributes.created_at);
+
+    if (connectorInfo.push != null && pushCreatedAt != null) {
+      pushRequest.push({ connectorId: connectorInfo.connectorId, date: pushCreatedAt });
+    }
+  }
+
+  if (pushRequest.length <= 0) {
+    return new Map();
+  }
+
+  const priorToPushFields = await userActionService.getConnectorFieldsUsedInPushes(
+    caseId,
+    pushRequest
+  );
+
+  const enrichedPushInfo = new Map<string, EnrichedPushInfo>();
+  for (const request of pushRequest) {
+    const connectorFieldsSO = priorToPushFields.get(request.connectorId);
+    const connectorFields = getConnectorInfoFromSavedObject(connectorFieldsSO);
+
+    if (connectorFields != null) {
+      enrichedPushInfo.set(request.connectorId, {
+        pushDate: request.date,
+        connectorFieldsUsedInPush: connectorFields,
+      });
+    }
+  }
+
+  return enrichedPushInfo;
+};
+
+const getDate = (timestamp: string | undefined): Date | undefined => {
+  if (timestamp == null) {
+    return;
+  }
+
+  const date = new Date(timestamp);
+
+  if (isDateValid(date)) {
+    return date;
+  }
+};
+
+const isDateValid = (date: Date): boolean => {
+  return !isNaN(date.getTime());
+};
+
+const getConnectorInfoFromSavedObject = (
+  savedObject: SavedObject<CaseUserActionResponse> | undefined
+): CaseConnector | undefined => {
+  if (
+    savedObject != null &&
+    (isConnectorUserAction(savedObject.attributes) ||
+      isCreateCaseUserAction(savedObject.attributes))
+  ) {
+    return savedObject.attributes.payload.connector;
+  }
+};
+
+const hasDataToPush = (enrichedConnectorInfo: EnrichedConnector): boolean => {
```

Comment:
The heart of the algorithm for determining if a connector should be flagged for needing to push again.